packages/core-data/src/reducer.js

```diff
@@ -127,26 +152,82 @@ export function themeSupports( state = {}, action ) {
  *
  * @return {Function} Reducer.
  */
-function entity( entityConfig ) {
-	return flowRight( [
-		// Limit to matching action type so we don't attempt to replace action on
-		// an unhandled action.
-		ifMatchingAction( ( action ) => (
-			action.name &&
-			action.kind &&
-			action.name === entityConfig.name &&
-			action.kind === entityConfig.kind
-		) ),
-
-		// Inject the entity config into the action.
-		replaceAction( ( action ) => {
+const data = ifMatchingEntityConfig( queriedDataReducer );
+
+const edits = ifMatchingEntityConfig( ( state = {}, action ) => {
+	switch ( action.type ) {
+		case 'EDIT_ENTITY_RECORD':
+			// TODO: When assigning an edit which is equal to the current data
+			// value, the edit should be omitted. This ensures that Undo to the
+			// last save point is treated as non-dirty.
 			return {
-				...action,
-				key: entityConfig.key || DEFAULT_ENTITY_KEY,
+				...state,
+				[ action.recordId ]: assignPropertyDescriptors(
+					{},
+					state[ action.recordId ],
+					action.edits,
+				),
 			};
-		} ),
-	] )( queriedDataReducer );
-}
+
+		case 'RECEIVE_ITEMS':
+			let nextState = state;
+			for ( const record of action.items ) {
+				const recordId = record[ action.key ];
+				if ( ! nextState.hasOwnProperty( recordId ) ) {
+					continue;
+				}
+
+				const nextEdits = reduce( nextState[ recordId ], ( result, value, key ) => {
+					// Deep equality avoids changing reference to a top-level
+					// key if all its properties match.
+					if ( ! isEqual( result[ key ], getRawValue( record[ key ] ) ) ) {
+						return result;
+					}
+
+					if ( result === nextState[ recordId ] ) {
+						result = assignPropertyDescriptors( {}, nextState[ recordId ] );
+					}
+
+					delete result[ key ];
+					return result;
+				}, nextState[ recordId ] );
+
+				if ( nextEdits === nextState[ recordId ] ) {
+					continue;
+				}
+
+				if ( nextState === state ) {
+					nextState = { ...state };
+				}
+
+				if ( Object.getOwnPropertyNames( nextEdits ).length ) {
+					nextState[ recordId ] = nextEdits;
+				} else {
+					delete nextState[ recordId ];
+				}
+			}
+
+			return nextState;
+	}
+
+	return state;
+} );
+
+const saving = ifMatchingEntityConfig( ( state = {}, action ) => {
```

Comment:
Curious why we need this. Can we use the resolving state?

Reply:
> Curious why we need this. Can we use the resolving state?

Unless I'm totally forgetting, that doesn't apply since resolvers are for selectors? Saving state occurs as a result of dispatching.

That being said, I could see some usefulness in knowing whether an action has "finished", and I imagine it would be possible to implement.

Reply:
>That being said, I could see some usefulness in knowing whether an action has "finished", and I imagine it would be possible to implement.

At the `@wordpress/data` level? That would be nice.